### ISSUE TITLE
Timeout for request issuing certificate

### ISSUE THREAD

**Issue by serathius** [GUI/lua-resty-auto-ssl#11]
First request that require server to issue a certificate is stuck on "Establishing secure connection".
After some time this shows in logs:
```
2016/07/23 15:26:24 [error] 9#9: lua tcp socket read timed out, context: ssl_certificate_by_lua*, client: 10.0.0.1, server: 0.0.0.0:443
2016/07/23 15:26:24 [error] 9#9: [lua] lets_encrypt.lua:20: issue_cert(): auto-ssl: letsencrypt.sh failed: env HOOK_SECRET=*** /usr/local/openresty/luajit/share/lua/5.1/resty/auto-ssl/vendor/letsencrypt.sh --cron --domain *** --challenge http-01 --config /etc/resty-auto-ssl/letsencrypt/config.sh --hook /usr/local/openresty/luajit/share/lua/5.1/resty/auto-ssl/shell/letsencrypt_hooks status: -1 out: nil err: timeout, context: ssl_certificate_by_lua*, client: 10.0.0.1, server: 0.0.0.0:443
2016/07/23 15:26:24 [error] 9#9: [lua] ssl_certificate.lua:70: issue_cert(): auto-ssl: issuing new certificate failed: letsencrypt.sh failure, context: ssl_certificate_by_lua*, client: 10.0.0.1, server: 0.0.0.0:443
2016/07/23 15:26:24 [error] 9#9: [lua] ssl_certificate.lua:238: ssl_certificate(): auto-ssl: could not get certificate for *** - using fallback - failed to get or issue certificate, context: ssl_certificate_by_lua*, client: 10.0.0.1, server: 0.0.0.0:443
```

The page shows bad certificate, but after reloading valid certificate is used.

I can't think of the secure way of avoiding exposing bad certificate, but would be nice if errors could be omitted.

**Comment by GUI** [comment]
Hm, the very first request should be handled without timing out or dropping the connection. The first request will be delayed slightly (while the certificate is being registered), but in my experience this usually doesn't last more than 2-5 seconds.

The only time I've seen these "timeout" errors is when I've been doing lots of local integration testing against Let's Encrypt's staging environment and I've been rate limited by Let's Encrypt. So one quick question is do you think you've tried registering enough domains in a short period of time to hit any Let's Encrypt [rate limits](https://letsencrypt.org/docs/rate-limits/)?

But you're case also sounds interesting, because you do get a valid certificate after the second request. So this leads me to believe that the certificate registration is succeeding, but perhaps just taking longer than we expect. After looking at the code, it does look like there is a [15 second default timeout](https://github.com/juce/lua-resty-shell/blob/0f88be3272c703686ef0d37f267f0616672c6931/lib/resty/shell.lua#L23) in one of the libraries we use that would explain this specific "timeout" error. If this 15 second timeout was hit, then I think the certificate registration could still eventually succeed in the background, but it would lead to this first failed request, like you describe.

So while I'm still not sure why the certificate registration is taking longer for you than normal, it could be a variety of things like rate limits, or temporary slowdowns on Let's Encrypt's end. So I think it could be worth us increasing our default timeout to at least allow for potential delays in a bit more graceful manner.

**Comment by GUI** [comment]
I've released v0.8.4 with an increase in the default timeout. I believe this should solve the issue you ran into, and generally make things better behaved if Let's Encrypt randomly takes longer sometimes.

But please let me know if you're still seeing any issues like this in v0.8.4.

Thanks for finding this!

**Comment by GUI** [comment]
@serathius: As a quick heads up, I recently uncovered the fact that this potential timeout issue wasn't correctly addressed in v0.8.4. Due to some rather silly oversights on my part, the default timeout had not in fact been extended from 15 seconds to 60 seconds. This meant there could still be potential problems if Let's Encrypt's servers took longer than 15 seconds to respond.

This bug should be properly addressed in v0.10.2, though, which I just published. In that version, the default timeout is now 60 seconds, and I'm pretty sure it's working this time. :)

Sorry for the mixup, but thanks again for the original report!

**Comment by pra-cloud** [comment]
My certificate renewal is not failed but i am getting this issue a lot and it makes my production down need help asap.

lua tcp socket read timed out, context: ssl_certificate_by_lua*, client: 172.69.58.204, server: 0.0.0.0:443

![image](https://github.com/auto-ssl/lua-resty-auto-ssl/assets/64471391/2f23a22a-8475-4fbd-a776-b601d7cefcd1)
